### PR TITLE
POST Firebase token and user data

### LINKED ISSUE #90: Fullstack: Implementing Firebase User Registration and Authentication
## User story

**Who:** **As a** developer
**What:** **I want to** be able to connect the Google Button to Firebase Auth service
**Why:** **so that we can** make sure that users can register through clicking on that button and signing in

## Implementation details

Use the existing code in src/client/firebase

https://firebase.google.com/docs/auth/web/google-signin
https://firebase.google.com/docs/auth/web/manage-users

And check the https://console.firebase.google.com/u/2/project/class16-fp-guide-it/ for more info

Once the token is generate you should add it to the user table on the DB

Before doing any development you should check this `src/client/firebase` and this https://github.com/HackYourFuture-`src/client/hooks` ( authentication hook) and this `src/client/components/Auth` (for protected routes)

### PR DESCRIPTION
**Pull request by NatsGt**
# Description

The FirebaseContext manages the states that the useAuthentication hook used to. The UserContext manages the user information passing it with useUser through all the App, and also sends the token and user information to the backend with fetch Post method. 

Fixes #90 

# How to test?

npm run dev http://localhost:3000/welcome

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have followed the name conventions for CSS Classnames and filenames, Components names and filenames, Style filenames, if you are in doubt check the the project README.MD and here https://github.com/HackYourFuture-CPH/curriculum/blob/master/review/review-checklist.md
- [x] I have commented my code, particularly in hard-to-understand areas, if you code was simple enough mark the box anyway
- [x] I have made corresponding changes to the documentation, if you code was simple enough mark the box anyway
- [x] This PR is ready to be merged and not breaking any other functionality

### CODE REVIEW

File: src/client/components/Auth/AuthenticatedRoute.js [hunk 1]
```diff
@@ -1,11 +1,11 @@
 import PropTypes from 'prop-types';
 import React from 'react';
-import { useAuthentication } from '../../hooks/useAuthentication';
 import { Redirect, Route } from 'react-router-dom';
 import Loader from '../Loader';
+import { useFirebase } from '../../firebase/FirebaseContext';
 
 function AuthenticatedRoute({ children, ...rest }) {
-  const { isAuthenticated, isLoading } = useAuthentication();
+  const { isAuthenticated, isLoading } = useFirebase();
```
Comment (senner007) [inline src/client/components/Auth/AuthenticatedRoute.js:8]: Great.
We need to replace useAuthentication with useFirebase everywhere. Not just here

File: src/server/api/controllers/boards.controller.js [hunk 2]
```diff
@@ -84,7 +84,15 @@ const getBoardsByCreatorId = async (id) => {
 
   const boards = await knex('users')
     .join('boards', 'users.id', '=', 'boards.creatorId')
```
Comment (senner007) [inline src/server/api/controllers/boards.controller.js:86]: This should not be in this pr

File: src/client/firebase/UserContext.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      const token = await getUserToken();
+      console.log(token, 'context');
+      const response = await fetch('/api/users', {
+        method: 'POST',
+        mode: 'cors',
+        headers: {
+          'Content-type': 'application/json',
+          authorization: `Bearer ${token}`,
+        },
+        body: JSON.stringify(authUser),
```
Comment (senner007) [inline src/client/firebase/UserContext.js:23]: we need to send the uid. So authUser.uid

**Review by senner007 — CHANGES_REQUESTED** [review]

**Comment by Midnighter** [comment]
Are you coordinating with @shanhaider23 on #173 because posting the token is basically creating the user, right?

File: src/client/firebase/UserContext.js (hunk 3 above)
Comment (NatsGt) [reply src/client/firebase/UserContext.js:23]: But then how would the person working on the backend get the information of the user? (name and photoURL)

Comment (senner007) [reply src/client/firebase/UserContext.js:23]: Correct yes, we need at least the name also. Can the photo url be extracted from the authUser?

Comment (NatsGt) [reply src/client/firebase/UserContext.js:23]: yes, name, email, uid, and photoURL. Should I send it as an object with this info or send the authUser as it is?

Comment (senner007) [reply src/client/firebase/UserContext.js:23]: 👌
Yes I see. I was actually wondering how to handle this. 
Better make an object then

Comment (senner007) [reply src/client/firebase/UserContext.js:23]: Then edit the users migration to add the photourl column

**Comment by NatsGt** [comment]
> Are you coordinating with @shanhaider23 on #173 because posting the token is basically creating the user, right?

yes, I'm passing information from the user so he can add it to the database.

**Review by senner007 — APPROVED** [review]
✔

**Review by senner007 — APPROVED** [review]
✔
Hi :)
I added a few changes. Can you verify and merge?

**Comment by NatsGt** [comment]
> ✔
> Hi :)
> I added a few changes. Can you verify and merge?

Everything is working fine 😃

**Pull request merged**